### ISSUE TITLE
Interaction use cases?

### ISSUE THREAD

**Issue by mmocny** [explainers-by-googlers/explicit-javascript-compile-hints-file-based#8]
I recently ran a TPAC breakout discussion on: https://github.com/mmocny/proposal-async-event-listeners

It touched on the topic of script loading event listeners, boostrapping/hydration effects, and/or lazy loading event handlers.  I see that this overlaps with Use case 2: interaction, which is listed in the explainer-- though I don't actually see any details about how that use case is meant to be addressed.

I wonder if there is any overlap here?

**Comment by marjakh** [comment]
The use case meant here is that part of the delay in user interactions is because of lazy function parsing & compilation. With explicit compile hints, we could avoid those delays by selecting code needed for certain (common / important) user interactions as eagerly compiled.

This feature only makes things more eager but never more lazy. So it can't be used for lazifying anything. If there's another mechanism for delaying something (like, delaying loading a JS file), you might use this feature for then making that thing eager when it's finally loaded.

I'm not an expert of the things you mentioned - do they contain JS functions that we'd like to make more eager than they are now?

**Comment by mmocny** [comment]
> This feature only makes things more eager but never more lazy.

Understood.  I was speculating about the potential to solve directly for making lazy things even more lazy.  Then, maybe everything else could be more eager by default.

However, I admit this is an unlikely change given the status quo and both efforts might be useful.

---

> I'm not an expert of the things you mentioned - do they contain JS functions that we'd like to make more eager than they are now?

The proposal(s) I linked are very nascent, but the patterns are common in existing frameworks.

The goal is to decrease the amount of JS loaded by default, but still support progressive enhancement, by allowing DOM nodes to express interest in event dispatch without actually loading and attaching a concrete event listener.  (Most components on any complex site will provide optional functionality which is not actually needed for any specific user visit).

Once a node is actually targeted for event dispatch via user interaction, now you would need to make that JS function exceptionally eager.

**Comment by marjakh** [comment]
Yes, definitely if the web moves to be "lazier", the browsers could start compiling more eagerly by default. However, I don't see that realistically happening, unfortunately.

You can also see this compile hint as a signal to the browser: "I have lazified my web page (delayed the loading of unnecessary stuff), you can handle the rest eagerly".
